### PR TITLE
Debug stream thread info pr

### PR DESCRIPTION
**Pull request by jsarha**
This PR contains following things:
- debug-stream definitions
- implementation for transporting the debug-stream records over debug window slot
- thread info implementation
  - extracts per thread stack and cpu usage statistics from Zephyr core
  - sends the statistics as debug-stream records
- python script for receiving the debug-stream records over debug window slot transport and decoding them to human readable form

This PR requires following PR to compile (if the feature is enabled in Kconfig):
https://github.com/zephyrproject-rtos/zephyr/pull/77467

and following PR to work (for python script to access the debug window slot):
https://github.com/thesofproject/linux/pull/5154

and this fix to be able to disable tlemetry performance measurements, that tries to use the same debug-window slot (a configuration that allows both is also possible, but I wanted to get this PR out):
https://github.com/thesofproject/sof/pull/9392

If one wants to try the PR in practice on MTL, one only needs the linux PR, and this branch of mine that contains all the other necessary changes:
https://github.com/jsarha/sof/tree/debug_stream_thread_info_testing

### CODE REVIEW

File: src/debug/debug_stream/debug_stream_slot.c [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+#include <adsp_memory.h>
+#include <sof/common.h>
+#include <rtos/string.h>
+#include "debug_stream.h"
+#include "debug_stream_slot.h"
+
+LOG_MODULE_REGISTER(debug_strem_slot);
+
+struct cpu_mutex {
+	struct k_mutex m;
```
Comment (lgirdwood) [inline src/debug/debug_stream/debug_stream_slot.c:17]: ok, I assume mutex is per core and not not lock for all core ? - if so we should inline comment this description.

File: src/debug/debug_stream/debug_stream_slot.c [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+int debug_stream_slot_send_record(struct debug_stream_record *rec)
+{
+	struct debug_stream_section_descriptor desc;
+	struct debug_stream_circular_buf *buf = get_circular_buffer(&desc, arch_proc_id());
+	uint32_t record_size = rec->size;
+	uint32_t record_start, buf_remain;
+
+	LOG_DBG("Sending record %u id %u len %u\n", rec->serial, rec->id, rec->size);
```
Comment (lgirdwood) [inline src/debug/debug_stream/debug_stream_slot.c:54]: Should this go after the buffer check below ?

File: src/debug/debug_stream/debug_stream_slot.c [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+		LOG_ERR("Record too big %u >= %u (desc %u %u %u)", rec->size,
+			desc.buf_words, desc.core_id, desc.buf_words, desc.offset);
+		return -ENOMEM;
+	}
+	k_mutex_lock(&cpu_mutex[arch_proc_id()].m, K_FOREVER);
+
+	rec->serial = buf->next_serial++;
+	rec->size = record_size + 1; /* +1 for size at the end of recored */
+	record_start = buf->w_ptr;
+	buf->w_ptr = (record_start + record_size) % desc.buf_words;
```
Comment (lgirdwood) [inline src/debug/debug_stream/debug_stream_slot.c:69]: modulus is expensive - will buffer words always be a power of 2 ?

File: src/debug/debug_stream/debug_stream_slot.h [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+ */
+
+#include <stdint.h>
+
+/* Core specific section descriptor
+ *
+ * Section descriptor defines core ID, offset and size of the circular
+ * buffer in the debug window slot.
+ */
+struct debug_stream_section_descriptor {
```
Comment (lgirdwood) [inline src/debug/debug_stream/debug_stream_slot.h:98]: again - is user API, needs to go in another directory and needs packed/aligned.

File: tools/debugstream/thread-info.py [hunk 5]
```diff
@@ -1,217 +0,0 @@
-#!/usr/bin/env python3
```
Comment (lgirdwood) [inline tools/debugstream/thread-info.py]: Does this need squashed into previous patch ?

**Review by marcinszkudlinski — CHANGES_REQUESTED** [review]

File: src/debug/debug_stream/Kconfig [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+config SOF_DEBUG_STREAM_SLOT
+	bool "Enable SOF debug stream debug window slot"
+	help
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/debug/debug_stream/Kconfig:5]: isn't it "n" by default?

File: tools/debugstream/thread-info.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+			    help='Telemetry2 window polling interval in seconds, default 1',
+                            default=1)
+	parser.add_argument('-f', '--telemetry2-file',
+			    help='File to read the telemetry2 data from, default /sys/kernel/debug/sof/telemetry2',
+                            default="/sys/kernel/debug/sof/telemetry2")
+	parsed_args = parser.parse_args()
+	return parsed_args
+
+args = parse_params()
+main_f(args)
```
Comment (lyakh) [inline tools/debugstream/thread-info.py]: this is supposed to run on the target system, right? So cross-OS compatibility isn't needed? So there should be no objections against having it in C / C++ ;-)

File: src/debug/debug_stream/debug_stream_thread_info.c [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+};
+
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+
+/* Finds and/or updates cached cycle counter value for 'tid'-thread
+ * and calculates used execution cycles since previous round and
+ * returns it.
+ */
+
+static uint32_t get_cycles(void *tid, k_thread_runtime_stats_t *thread_stats,
```
Comment (lyakh) [inline src/debug/debug_stream/debug_stream_thread_info.c:87]: namespace prefixes make searching easier

File: src/debug/debug_stream/debug_stream_thread_info.c [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+				   THREAD_STACK_SIZE);
+static struct k_thread info_thread[CONFIG_MP_MAX_NUM_CPUS];
+
+static int thread_info_start(void)
+{
+	uint32_t i;
+
+	for (i = 0; i < ARRAY_SIZE(info_thread); i++) {
+		char name[24];
+		k_tid_t tid = NULL;
```
Comment (lyakh) [inline src/debug/debug_stream/debug_stream_thread_info.c:357]: superfluous init

File: tools/debugstream/debug-stream.py [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+		if record != None:
+			success = self.printer.print_record(record, self.cpu)
+			if success:
+				recp = ctypes.cast(record, ctypes.POINTER(DebugStreamRecord))
+				self.prev_w_ptr = (self.prev_w_ptr + recp.contents.size
+						   ) % self.buf_words
+				self.prev_serial = recp.contents.serial
+			else:
+				logging.info(f"Parse failed on record {self.prev_serial + 1}")
+			return success
```
Comment (kv2019i) [inline tools/debugstream/debug-stream.py:259]: Success even if "Parse failed on record" ? Looks at least a bit odd

File: tools/debugstream/debug-stream.py [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+class CircularBufferDecoder:
+	desc = None
+	boffset = None
+	buf_words = None
+	cpu = None
+	printer = None
+	prev_w_ptr = 0
+	prev_serial = None
+	error_count = 0
+	def __init__(self, desc, cpu, printer):
```
Comment (kv2019i) [inline tools/debugstream/debug-stream.py:143]: Whitespace before first function would be nice.

File: src/debug/debug_stream/debug_stream_slot.c [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+	buf_remain = desc.buf_words - record_start;
+	if (buf_remain < record_size) {
+		uint32_t rec_remain = record_size - buf_remain;
+
+		assert(!memcpy_s(&buf->data[record_start], buf_remain * sizeof(uint32_t),
```
Comment (kv2019i) [inline src/debug/debug_stream/debug_stream_slot.c:74]: Please no code in asserts. If you build with asserts disabled, the functionality will not be the same.

**Review by kv2019i — CHANGES_REQUESTED** [review]
No showstoppers, aside the assert usage.

File: tools/debugstream/debug-stream.py [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+
+""" Generic Debug-stream header """
+class DebugStreamHdr(ctypes.Structure):
+	_fields_ = [
+		("magic", ctypes.c_uint),
+		("hdr_size", ctypes.c_uint),
+	]
+
+""" Debug Stream record for passing debug data """
+class DebugStreamRecord(ctypes.Structure):
```
Comment (marc-hb) [inline tools/debugstream/debug-stream.py:26]: This is not a valid docstring, it should be on the next line:
```
pylint debugstream/debug-stream.py

tools/debugstream/debug-stream.py:25:0: W0105: String statement has no effect (pointless-string-statement)
tools/debugstream/debug-stream.py:26:0: C0115: Missing class docstring (missing-class-docstring)
```

**Review by marc-hb — DISMISSED** [review]
`debug-stream.py` has 352 `pylint` warnings: that's almost as many as the number of lines in the file! I'm NOT requesting "pylint perfection" but the number of warnings should be small enough to keep `pylint` usable. It's not usable right now.

The good news is: 256 out of these 352 are the same W0311 and you can get rid of ALL of them with a single `black` call:

```
apt install black
black tools/debugstream/debug-stream.py
reformatted tools/debugstream/debug-stream.py
```
Warning: `black` reformatting happens _in place_.

Another pro tip: `sorted_pylint () {     pylint "$@" | sort -t: -k2 -n ; }`

```
tools/debugstream/debug-stream.py:370:0: W0311: Bad indentation. Found 5 spaces, expected 20 (bad-indentation)
tools/debugstream/debug-stream.py:371:0: W0311: Bad indentation. Found 5 spaces, expected 20 (bad-indentation)
tools/debugstream/debug-stream.py:372:0: W0311: Bad indentation. Found 6 spaces, expected 24 (bad-indentation)
tools/debugstream/debug-stream.py:374:0: W0311: Bad indentation. Found 2 spaces, expected 8 (bad-indentation)
tools/debugstream/debug-stream.py:375:0: W0311: Bad indentation. Found 3 spaces, expected 12 (bad-indentation)
tools/debugstream/debug-stream.py:376:0: W0311: Bad indentation. Found 3 spaces, expected 12 (bad-indentation)
tools/debugstream/debug-stream.py:377:0: W0311: Bad indentation. Found 2 spaces, expected 8 (bad-indentation)
tools/debugstream/debug-stream.py:378:0: W0311: Bad indentation. Found 3 spaces, expected 12 (bad-indentation)
tools/debugstream/debug-stream.py:379:0: W0311: Bad indentation. Found 4 spaces, expected 16 (bad-indentation)
tools/debugstream/debug-stream.py:380:0: W0311: Bad indentation. Found 4 spaces, expected 16 (bad-indentation)
tools/debugstream/debug-stream.py:381:0: W0311: Bad indentation. Found 2 spaces, expected 8 (bad-indentation)
tools/debugstream/debug-stream.py:382:0: W0311: Bad indentation. Found 2 spaces, expected 8 (bad-indentation)
```

**Comment by marc-hb** [comment]
> but the number of warnings should be small enough to keep pylint usable. 

Forgot this sorry: most warnings are easy to fix but in some rare cases pylint is wrong or you don't have the time to figure out how (feel free to ask for help privately). Then you can easily silence any warning with a `# pylint: disable...` comment. Either in the spot (preferred), or at in the whole file. Find examples in `git -C sof[-test] grep pylint.*disable`.

Or you can just leave some for someone else to fix later.

File: src/debug/debug_stream/Kconfig [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+if SOF_DEBUG_STREAM_SLOT
+
+config SOF_DEBUG_STREAM_SLOT_NUMBER
+	int "Debug window slot where to put debug stream slot"
+	default 3
+	range 0 14
+	help
+	  Which debug slot to reserve for Debug Stream. Remember to map
+	  the slot with MEMORY_WIN_2_SIZE in soc/intel/intel_adsp/Kconfig,
+	  in Zephyr source tree.
```
[the remark this answers is not in the document]
Comment (jsarha) [reply src/debug/debug_stream/Kconfig:21]: The necessary Zephyr features are selected in a following commit.

File: src/debug/debug_stream/debug_stream_slot.c (hunk 2 above)
Comment (jsarha) [reply src/debug/debug_stream/debug_stream_slot.c:54]: A matter or taste. I wanted to know for sure the function is at least called, when I could not see anything happening in the circular buffer.

File: src/debug/debug_stream/debug_stream_slot.c (hunk 3 above)
Comment (jsarha) [reply src/debug/debug_stream/debug_stream_slot.c:69]: @lgirdwood , No its not. Forcing it to be power of 2 could potentially waste half of the 4k slot. Not sure if I should optimize the slot usage or execution time. In any case the memcpy() to uncached memory is probably the most expensive operation here, not the single modulus operation. In any case the records are not sent very often often, now five messages in 10 seconds. Should not overload the DSP.

File: src/debug/debug_stream/debug_stream_slot.h (hunk 4 above)
Comment (jsarha) [reply src/debug/debug_stream/debug_stream_slot.h:98]: Packed hardly does anything when everything is 32-bit words, and the descriptor is already aligned to cache-line size. The header itself must be written to beginning of slot so its automatically 4k page size aligned, which should be enough, and also if its not the idea is still to have it at the  very beginning of the slot. I can put the packed aligned stuff to the header too, its not needed, and I feel it could even be misleading.

File: tools/debugstream/thread-info.py (hunk 5 above)
Comment (jsarha) [reply tools/debugstream/thread-info.py]: Actually it looks like I accidentally squashed some py-script-commit to an earlier patch... will fix.

File: src/debug/debug_stream/debug_stream_slot.c [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+LOG_MODULE_REGISTER(debug_strem_slot);
+
+struct cpu_mutex {
+	struct k_mutex m;
+} __aligned(CONFIG_DCACHE_LINE_SIZE);
+
+static struct cpu_mutex cpu_mutex[CONFIG_MP_MAX_NUM_CPUS];
+
+static const int slot_num = CONFIG_SOF_DEBUG_STREAM_SLOT_NUMBER;
```
[the remark this answers is not in the document]
Comment (jsarha) [reply src/debug/debug_stream/debug_stream_slot.c:22]: Sure, will do.

File: src/debug/debug_stream/debug_stream_slot.c [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+static struct debug_stream_slot_hdr *get_my_slot(void)
```
[the remark this answers is not in the document]
Comment (jsarha) [reply src/debug/debug_stream/debug_stream_slot.c:24]: Sure.

File: tools/debugstream/thread-info.py (hunk 7 above)
Comment (jsarha) [reply tools/debugstream/thread-info.py]: I was explicitly advised to use python, I am not going to change it to C now.

File: src/debug/debug_stream/debug_stream_thread_info.c [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+		ret = k_thread_cpu_pin(tid, i);
+		if (ret < 0) {
+			LOG_ERR("Pinning thread to code core %u", i);
+			k_thread_abort(tid);
+			continue;
+		}
+		snprintf(name, sizeof(name), "%u thread info", i);
+		ret = k_thread_name_set(tid, name);
+		if (ret < 0)
+			LOG_INF("k_thread_name_set failed: %d for %u", ret, i);
```
[the remark this answers is not in the document]
Comment (jsarha) [reply src/debug/debug_stream/debug_stream_thread_info.c:378]: There is only two reasons why k_thread_name_set() would fail. Either the  thread is not found (should not happen here), or CONFIG_THREAD_NAME=n. So this is mostly just a gentle indication that the thread names are configured.

File: src/debug/debug_stream/debug_stream_slot.c (hunk 3 above)
Comment (lgirdwood) [reply src/debug/debug_stream/debug_stream_slot.c:69]: ok, so no power of 2, we can live with % (which really is a CC builtin func that loops on xtensa)

File: src/debug/debug_stream/debug_stream_slot.h (hunk 4 above)
Comment (lgirdwood) [reply src/debug/debug_stream/debug_stream_slot.h:98]: best to be explicit here even though parent is packed and this is all 32bit words, i.e. we could have direct users that may reference this structure without the parent context (and hence optimize for 32 but aligned/packed access).

File: src/debug/debug_stream/Kconfig (hunk 6 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/debug/debug_stream/Kconfig:5]: Lets call it out though, we really should have a default 'n' or 'y' for all Kconfig entries.

File: src/debug/debug_stream/debug_stream.h [hunk 18]
```diff
@@ -0,0 +1,52 @@
+/* SPDX-License-Identifier: BSD-3-Clause
```
[the remark this answers is not in the document]
Comment (jsarha) [reply src/debug/debug_stream/debug_stream.h:1]: Will do. I'll also split the structs defining the thread-info records into a separate header and I'll put that to src/include/user too.

File: src/debug/debug_stream/debug_stream.h [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+ * debug entity to user space debug tool for decoding and
+ * presentation. The data is recognized by the 'id' and the 'size'
+ * describes the amount of data. The 'serial' is a running number of sent
+ * record, increased by one after each record. The protocol is
+ * agnostic about the contents of the records.
+ */
+
+struct debug_stream_record {
+	uint32_t id;		/* Record id of abstract data record */
+	uint32_t serial;	/* Increments after each record */
```
[the remark this answers is not in the document]
Comment (jsarha) [reply src/debug/debug_stream/debug_stream.h:43]: I'll make another version with everything else fixed first at least.

File: src/debug/debug_stream/Kconfig (hunk 6 above)
[the remark this answers is not in the document]
Comment (marc-hb) [reply src/debug/debug_stream/Kconfig:5]: The Zephyr coding style recommends to avoid the redundant `default n` for a couple reasons:

https://docs.zephyrproject.org/latest/build/kconfig/tips.html#redundant-defaults

`git -C zephyr/ grep -h default.n -- '*Kconfig' | wc -l` is 10 times smaller than `default y` so it seems mostly followed in Zephyr.

The same SOF stats look very different indeed.

File: src/debug/debug_stream/debug_stream_slot.c (hunk 3 above)
Comment (lyakh) [reply src/debug/debug_stream/debug_stream_slot.c:69]: @lgirdwood I think on our recent Xtensa chips `QUOS` is used for division, also by the Zephyr toolchain

File: src/debug/debug_stream/debug_stream_slot.h (hunk 4 above)
Comment (lyakh) [reply src/debug/debug_stream/debug_stream_slot.h:98]: @jsarha FWIW agree with @lgirdwood - please add `packed`. It's good for all ABI types to have it even if they "obviously" effectively would be packed by default by the compiler anyway.

Comment (jsarha) [reply src/debug/debug_stream/debug_stream_slot.h:98]: It looks like if in xcc compilation (I have not tried with gcc) I put __packed to a struct, that contains an array of structs that are __aligned(64), then the first array elemen is not 64-byte aligined, but packed next to previous element, second is then 64-bytes after the first, but still wrongly aligned. 

So I can not make struct debug_stream_slot_hdr __packed, because then the elements of section_desc[] are not 64-byte aligned. And adding __aligned(64) to   too does not help. So at least I have to drop the __packed from struct debug_stream_slot_hdr.

Comment (marc-hb) [reply src/debug/debug_stream/debug_stream_slot.h:98]: > if put __packed to a struct, that contains an array of structs that are __aligned(64), then the first array elemen is not 64-byte aligined, but packed next to previous element, second is then 64-bytes after the first, but still wrongly aligned.

Can provide the C code and `pahole` output? I'm not 100% sure I follow. How can the first element have a "previous" element?

I suspect `aligned(64)` could be automatically downgraded to `aligned(32)` in some 32 bits environments. Quoting https://gcc.gnu.org/onlinedocs/gcc/Common-Type-Attributes.html#index-packed-type-attribute

> On many systems, the linker is only able to arrange for variables to be aligned up to a certain maximum alignment. [...] If your linker is only able to align variables up to a maximum of 8-byte alignment, then specifying aligned (16) in an `__attribute__` still only provides you with 8-byte alignment.

Comment (lyakh) [reply src/debug/debug_stream/debug_stream_slot.h:98]: @jsarha alignment doesn't always play well with packing. You pack data explicitly to make it "predictable," "reproducible." And it *can well* break alignment. Or at least you take care of alignment yourself. E.g. if you have
```
struct x {
    uint8_t c;
    uint32_t i;
} __attribute__((packed));
```
then sure enough `i` will be misaligned, and if you attempt to access it directly on some architectures you get an alignment violation exception. So you either add padding there or rearrange members manually to achieve desired alignment. So, I'd say that ABI types should be packed and alignment should be provided manually, or you end up using `memcpy()` to access unaligned members.

Comment (marc-hb) [reply src/debug/debug_stream/debug_stream_slot.h:98]: > And it can well break alignment

I think you mean: break "normal" alignment.

I'm relatively confident `packed` is in most cases equivalent to `packed,aligned(1)` where `packed,aligned(X)` means: align on exactly X while `aligned(X)` (not packed) means: aligned on _at least_ X but let the optimizer align higher than X if it thinks it knows better.


> I suspect aligned(64) could be automatically downgraded to aligned(32) in some 32 bits environments.

A clearer statement here: https://www.gnu.org/software/c-intro-and-ref/manual/html_node/Attributes.html
> However, you can’t force an alignment bigger than the computer’s maximum meaningful alignment.

Comment (jsarha) [reply src/debug/debug_stream/debug_stream_slot.h:98]: I can of course create all kinds of constructs to invent the __aligned() again with variable sized paddings, and use __packed, but now I do not have time for it so, I leave this as it is.

File: src/debug/debug_stream/Kconfig [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+config SOF_DEBUG_STREAM_THREAD_INFO
+	bool "Enable Zephyr thread info reporting through Debug-Stream"
+	select INIT_STACKS
+	select THREAD_MONITOR
+	select THREAD_STACK_INFO
+	select THREAD_RUNTIME_STATS
```
[the remark this answers is not in the document]
Comment (jsarha) [reply src/debug/debug_stream/Kconfig:28]: Those options are directly copied from here:
https://docs.zephyrproject.org/2.7.5/reference/kconfig/CONFIG_THREAD_ANALYZER.html

File: src/debug/debug_stream/debug_stream_slot.h (hunk 4 above)
Comment (lyakh) [reply src/debug/debug_stream/debug_stream_slot.h:98]: @jsarha Wouldn't just removing `__aligned` and using `__packed` already solve it? And just look through those types to verify that you arrange elements from larger to smaller - first all 64-bit ones, then 32-bit ones etc. Only if you embed structures, well, then yes, just pad the structure that you're embedding with some `uint8_t reserved[N]` array to align to the next 64 bits, that's it

Comment (jsarha) [reply src/debug/debug_stream/debug_stream_slot.h:98]: Yes, there is several ways to do this, and probably it would work even if the core specific  circular-buffer descriptors were not 64-byte aligned, since after the initialization the contents is read only and the memory is accessed through uncached address-space. Still @lgirdwood suggested that also the static descriptor part should be 64-bit aligned, just to be safe. I am really fine with anything that works, but I would like to do it only once more, and the time is running little short if we want this to go to 2.11.

File: tools/debugstream/debug-stream.py [hunk 21]
```diff
@@ -0,0 +1,399 @@
...
+import sys
+
+import logging
+logging.basicConfig(format="%(filename)s:%(lineno)s %(funcName)s: %(message)s",
+		    level=logging.WARNING)
+
+DEBUG_STREAM_PAYLOAD_MAGIC = 0x1ED15EED
+
+""" Generic Debug-stream header """
+class DebugStreamHdr(ctypes.Structure):
```
[the remark this answers is not in the document]
Comment (marc-hb) [reply tools/debugstream/debug-stream.py:19]: Sounds fair, maybe just add a "TODO: check if construct can replace this?" so that information does not get lost?

File: tools/debugstream/debug-stream.py [hunk 22]
```diff
@@ -0,0 +1,399 @@
...
+		try:
+			with open(args.debugstream_file, "rb") as f:
+				decoder.set_file(f)
+				decoder.update_slot()
+				if not decoder.get_descriptors():
+					break
+				decoder.catch_up_all()
+				while True:
+					if decoder.poll():
+						time.sleep(args.update_interval)
```
[the remark this answers is not in the document]
Comment (jsarha) [reply tools/debugstream/debug-stream.py:369]: The best document is in the second commit in debug_stream_slot.h.

[the remark this answers is not in the document]
Comment (marc-hb) [reply tools/debugstream/debug-stream.py]: That looks like great documentation but I didn't find why the design requires polling. I spent very little time reading either so I'll just trust you :-)

**Comment by jsarha** [comment]
The new version should address all the comment according to reply, except for "serial" > "seqnum" and getting rid off __aligned() and using __packed and padding members instead.

**Comment by jsarha** [comment]
Now also "serial" -> "seqnum" change is done. Aome idle debug prints were also removed.

File: src/debug/debug_stream/debug_stream_thread_info.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (kv2019i) [reply src/debug/debug_stream/debug_stream_thread_info.c:362]: Ack

**Review by kv2019i — APPROVED** [review]

File: src/debug/debug_stream/debug_stream_slot.h (hunk 4 above)
Comment (jsarha) [reply src/debug/debug_stream/debug_stream_slot.h]: @lgirdwood @lyakh , I am more and more starting to thing the __aligned() should be dropped from "struct debug_stream_section_descriptor", and for that matter all the padding hacks too. We can not have the ABI depend on cache-line size, or at least we should then write the offsets explicitly out in the debug window, but that is pretty much what the "struct debug_stream_section_descriptor" is for:  It is all static data that is anyway initialized by core 0 before the threads running on other cores have started and its in uncached memory, and it contains the offsets of the core specific circular buffers, and their sizes. So, there should not and must not be any cash-line size dependent padding in the header that describes where the core-specific circular buffers are.  I've tested this works fine with sof-mtl-nocodec-ssp0-ssp2.tplg on MTL.

File: src/debug/debug_stream/debug_stream_slot.c [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+		assert(!ret);
+	} else {
+		int ret;
+
+		ret = memcpy_s(&buf->data[record_start], buf_remain * sizeof(uint32_t),
+			       rec, record_size * sizeof(uint32_t));
+		assert(!ret);
+	}
+	/* Write record size again after the record */
+	buf->data[buf->w_ptr] = record_size + 1;
```
Comment (lyakh) [inline src/debug/debug_stream/debug_stream_slot.c:91]: that writes the size of the previous record after it? Why is this needed?

File: src/debug/debug_stream/debug_stream_thread_info.c [hunk 24]
```diff
@@ -0,0 +1,371 @@
...
+ */
+
+static uint32_t thread_info_get_cycles(void *tid, k_thread_runtime_stats_t *thread_stats,
+				       struct user_data *ud, const char *name)
+
+{
+	int i;
+
+	/* Mark the thread as active */
+	ud->active_threads[ud->thread_count] = tid;
```
Comment (lyakh) [inline src/debug/debug_stream/debug_stream_thread_info.c:78]: `thread_count` sounds like it's the number of threads, i.e. it begins with 1? Can this access outside of the array?

File: src/debug/debug_stream/debug_stream_thread_info.c [hunk 25]
```diff
@@ -0,0 +1,371 @@
...
+};
+
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+
+/* Finds and/or updates cached cycle counter value for 'tid'-thread
+ * and calculates used execution cycles since previous round and
+ * returns it.
+ */
+
+static uint32_t thread_info_get_cycles(void *tid, k_thread_runtime_stats_t *thread_stats,
```
Comment (lyakh) [inline src/debug/debug_stream/debug_stream_thread_info.c:71]: can we use `struct k_thread *`?

File: src/debug/debug_stream/debug_stream_thread_info.c [hunk 26]
```diff
@@ -0,0 +1,371 @@
...
+			ud->previous->threads[i].cycles = thread_stats->execution_cycles;
+			LOG_DBG("%p placed at %d (%s %llu)", tid, i,
+				name, ud->previous->threads[i].cycles);
+			break;
+		}
+	}
+
+	/* If there is more than THREAD_INFO_MAX_THREADS threads in this core */
+	if (i == ARRAY_SIZE(ud->previous->threads))
+		LOG_INF("No place found for %s %p", name, tid);
```
Comment (lyakh) [inline src/debug/debug_stream/debug_stream_thread_info.c:107]: warning?

File: src/debug/debug_stream/debug_stream_thread_info.c [hunk 27]
```diff
@@ -0,0 +1,371 @@
...
+			/* update cached value */
+			ud->previous->threads[i].cycles = thread_stats->execution_cycles;
+			return cycles;
+		}
+	}
+
+	/* If no cached value was found, look for an empty slot to store the recent value */
+	for (i = 0; i < ARRAY_SIZE(ud->previous->threads); i++) {
+		if (ud->previous->threads[i].tid == NULL) {
+			ud->previous->threads[i].tid = tid;
```
Comment (lyakh) [inline src/debug/debug_stream/debug_stream_thread_info.c:97]: can be a follow-up, but maybe we can have a better name than `previous` here. You don't store more than one snapshot, so isn't it just "current"?

File: src/debug/debug_stream/debug_stream_thread_info.c [hunk 28]
```diff
@@ -0,0 +1,371 @@
...
+	for (;;) {
+		thread_info_get(core, &rec);
+		k_sleep(K_SECONDS(CONFIG_SOF_DEBUG_STREAM_THREAD_INFO_INTERVAL));
+	}
+}
+
+#define THREAD_STACK_SIZE (2048)
+static K_THREAD_STACK_ARRAY_DEFINE(info_thread_stacks, CONFIG_MP_MAX_NUM_CPUS,
+				   THREAD_STACK_SIZE);
+static struct k_thread info_thread[CONFIG_MP_MAX_NUM_CPUS];
```
Comment (lyakh) [inline src/debug/debug_stream/debug_stream_thread_info.c:333]: `CONFIG_MP_NUM_CPUS`?

**Review by lyakh — COMMENTED** [review]
As long as none of my questions actually uncover bugs and since this is new code, they can be addressed after merge, nothing seems to be critical

File: src/debug/debug_stream/debug_stream_thread_info.c (hunk 28 above)
Comment (jsarha) [reply src/debug/debug_stream/debug_stream_thread_info.c:333]: Oh, that is obsolete. Will change to MP_MAX_NUM_CPUS Zephyr option, unles you now some better way to know number of CPUs we should prepare for at compile time. 

config MP_MAX_NUM_CPUS
        int "Maximum number of CPUs/cores"
        default 1
        range 1 12
        help
          Maximum number of multiprocessing-capable cores available to the
          multicpu API and SMP features.

File: src/debug/debug_stream/debug_stream_thread_info.c (hunk 27 above)
Comment (jsarha) [reply src/debug/debug_stream/debug_stream_thread_info.c:97]: It the cycles number from the previous round. After the calculations it will be "current", but when its used next time its again from "previous" round.

**Comment by jsarha** [comment]
Following changes were made:
- No cache-line size alingment for static data circular buffer descriptors
- record member "size" to "size_words"
- struct record_buf called bufd and rec to avoid confusion
- Typo-fixes
- TODO comment for using python construct instead of ctypes stucts for data models
- thread_count overflow check
- INF to WRN logging promotion for thread_count overflow
- Comment out sys_cache_data_flush_range()
- Add comment for using "void *tid" in thread_info_get_cycles()

File: src/debug/debug_stream/debug_stream_slot.c [hunk 29]
```diff
@@ -0,0 +1,150 @@
...
+#include <user/debug_stream_slot.h>
+
+LOG_MODULE_REGISTER(debug_strem_slot);
+
+struct cpu_mutex {
+	struct k_mutex m;
+} __aligned(CONFIG_DCACHE_LINE_SIZE);
+
+/* CPU specific mutexes for each circular buffer */
+static struct cpu_mutex cpu_mutex[CONFIG_MP_MAX_NUM_CPUS];
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/debug/debug_stream/debug_stream_slot.c:21]: @jsarha I'd keep them uncached and not bother with cache lines. Should be both simpler and safer. Though a bit slower, but don't think this slow down would be significant for you. But if you do decide to go via cache, you can get cached aliases to that memory, but then yes, you'd need to reserve a whole cache line-sized buffer for each.

File: src/debug/debug_stream/debug_stream_slot.c [hunk 30]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS; i++) {
+		struct debug_stream_section_descriptor desc;
+		struct debug_stream_circular_buf *buf =
+			debug_stream_get_circular_buffer(&desc, i);
+
+		buf->next_seqno = 0;
+		buf->w_ptr = 0;
+		k_mutex_init(&cpu_mutex[i].m);
+		/* The core specific mutexes are now .dss which is uncached so the
```
Comment (lyakh) [inline src/debug/debug_stream/debug_stream_slot.c:143]: `.bss`

File: src/debug/debug_stream/debug_stream_thread_info.c (hunk 24 above)
Comment (lyakh) [reply src/debug/debug_stream/debug_stream_thread_info.c:85]: @jsarha but does it begin with 0 or with 1? I.e. does the element `ud->active_threads[0]` ever get used here?

**Review by lyakh — APPROVED** [review]
thanks for addressing comments

File: src/debug/debug_stream/debug_stream_slot.h (hunk 4 above)
Comment (lyakh) [reply src/debug/debug_stream/debug_stream_slot.h]: @jsarha I agree. The padding that I was proposing wasn't for cache-line size alignment, it was for "natural data type alignment." I.e. if you have an 8-bit member followed by a 16-bit value and you cannot swap them, add a padding like
```
struct abi_data {
    uint8_t type;
    uint8_t reserved0[3];
    uint32_t size;
    ....
};
```

**Comment by kv2019i** [comment]
@marcinszkudlinski  ok for you to proceed? default for this is "n".

**Review by marcinszkudlinski — APPROVED** [review]

**Comment by kv2019i** [comment]
@wszypelt  Can you check? This seems to have failed to "Copy build to DUT" step.

**Comment by jsarha** [comment]
> @jsarha but does it begin with 0 or with 1? I.e. does the element ud->active_threads[0] ever get used here?

@lyakh from zero of course, like in every decent c-program:
 
https://github.com/jsarha/sof/blob/5fe574452db992cdda5ae5cdcfc560adc0862102/src/debug/debug_stream/debug_stream_thread_info.c#L275

https://github.com/jsarha/sof/blob/5fe574452db992cdda5ae5cdcfc560adc0862102/src/debug/debug_stream/debug_stream_thread_info.c#L229
